src/Core/Content/Category/SalesChannel/NavigationRoute.php

```diff
@@ -197,13 +198,18 @@ private function getMissingIds(string $activeId, ?string $path, array $childIds,
         return array_values(array_diff($haveToBeIncluded, $included));
     }
 
-    private function validate(string $activeId, ?string $path, SalesChannelContext $context): void
+    protected function getValidSalesChannelIds(SalesChannelEntity $salesChannel): array
```

Comment:
This change still would not allow for proper decoration or? As the method is only protected and not part of the abstract class, that means a decorator can not call that method from the inner class, meaning the only way to change this behaviour would be to completely override that class instead of properly decorating it, which in turn comes with issues like multiple plugins overriding each other etc

to be able to properly decorate that functionality i guess that new method should be a new public method on the abstract class, that way you could properly decorate the route in a way that is compatible with mutliple plugins using that extension point as well:

```php
class MyDecoratedNavigationRoute extends AbstractNavigationRoute

    public function __construct(private AbstractNavigationRoute $inner)
    {}
    
    public function getValidSalesChannelIds(SalesChannelEntity $salesChannel): array
    {
        $ids = $this->inner->getValidSalesChannelIds($salesChannel);
        
        $ids[] = $mySpecialId;
        
        return $ids;
}
```

Also thinking about this i guess we can be more precise in the naming as the method won't return ids for saleschannels but rather categories from this sales channel. so i would go with something like `getValidCategoryIdsForSalesChannel($salesChannel)`

What do you think?

Reply:
maybe an event is enough here so you can modifiy the ids? 🤔

Reply:
@keulinho you're right. Didn't think about this part yet. I have created a `NavigationRouteValidateEvent` as @shyim suggested. Also I replace some function of the NavigationRoute class to a generic `CategoryService` so this code can be reused and the code is a lot cleaner.

Reply:
should we use our new extension system? :thinking: 

https://github.com/shopware/shopware/blob/trunk/adr/2024-06-18-extended-event-system.md

Reply:
I appreciate your guidance and would like to understand the reasoning behind this question better. Is there a place where I can read more about the new extension system instead of the given link? I would love to read about the main differences between the new extension system and the standard Symfony event system. My main question is when you should choose the extension system over the event system.